From patchwork Mon Apr 15 13:10:16 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Paolo Bonzini <pbonzini@redhat.com>
X-Patchwork-Id: 10900827
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id DE62D1390
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Apr 2019 13:10:22 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id C7B4F2842B
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Apr 2019 13:10:22 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id BB8DF28872; Mon, 15 Apr 2019 13:10:22 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.7 required=2.0 tests=BAYES_00,DKIM_INVALID,
	DKIM_SIGNED,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 68B752842B
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Apr 2019 13:10:22 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727445AbfDONKU (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 15 Apr 2019 09:10:20 -0400
Received: from mail-wr1-f66.google.com ([209.85.221.66]:39783 "EHLO
        mail-wr1-f66.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727316AbfDONKU (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 15 Apr 2019 09:10:20 -0400
Received: by mail-wr1-f66.google.com with SMTP id j9so21867638wrn.6
        for <kvm@vger.kernel.org>; Mon, 15 Apr 2019 06:10:19 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=sender:from:to:subject:date:message-id;
        bh=i+v2pvFOC8oWHN5LqwgzsZUwaEHNf+K28JfzgCrtaHQ=;
        b=lx2z4bMJfEPW84Maz392gOkuUmyR86R96RDSM1+ZLcOrkGPwXFb3dwnvGopM7ne/h5
         vHUg+QgRV1OKi7i/202E9f8iY4vrvl4W6e/hhsZGjdzJcZ7IylUqnWtbyRRHgFuFoJEM
         OnxMAMWMq9wticot4cgPxFoHPrCNDWXB0NU+LXnzTs+4ekes5cL4OwKxLoHuGidMnsFf
         jGZZ0G/WrCcPYAvCCePZTFAvQ/RRhAmDpDqUbb842c/7wupFCmC/sNc9717Kjzm9snqx
         UU72mpiM1fbkIiyGCb527gFQeBIS1/h0ALNRR93cYOlEmnAosq0jwUUK+d7otxr51KKm
         AG6g==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:sender:from:to:subject:date:message-id;
        bh=i+v2pvFOC8oWHN5LqwgzsZUwaEHNf+K28JfzgCrtaHQ=;
        b=WY2XzbG4nsASLM+j1AND0iUJuBGa6dAbrQMU4lRF3+/9nVaeT71zncOG1Anj5uwBEi
         pg2sz+/O+8I/jHgWuO1P+q2AoNLMHytA3CYdAMCZlGRnzzIg6BkKoNOevV0o+r3kDfQe
         AHLUmidcvM5SvFU9hKaf0l5Tkbkae6eL3L+kYQY2nBFEtyFBuZuAXJKcsGQEyIwlvwTy
         3rOJFcfSOtBqBWeB1vEeD+AMqVxtwNlaUN7+o6vmaH0EPHTDcSqNhSvGoOEHn+kHtNff
         DC9Is7gYNVGQgaGCL2fyR7Ncb9bg9NyORoE14/2jEtlkZQSV7+kyBVUHCB0N1WNf0uq9
         GJhg==
X-Gm-Message-State: APjAAAV1q03xBJ+DFhseXoHfLhON0trdgA/NZ1KqE/Mz+QL0HLCU31ga
        CDbfz+FqwTHGnVurOfHHYIFMeGnS
X-Google-Smtp-Source: 
 APXvYqwXA158/tpsCVl19GA/jEKI9gpfliDIgU5IXdwAbKlILAoDXjVfZm13fR/TVAznNUUqjrHc7w==
X-Received: by 2002:adf:eb02:: with SMTP id s2mr51087637wrn.29.1555333818524;
        Mon, 15 Apr 2019 06:10:18 -0700 (PDT)
Received: from 640k.lan ([93.56.166.5])
        by smtp.gmail.com with ESMTPSA id
 c6sm17667149wmb.21.2019.04.15.06.10.17
        for <kvm@vger.kernel.org>
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Mon, 15 Apr 2019 06:10:17 -0700 (PDT)
From: Paolo Bonzini <pbonzini@redhat.com>
To: kvm@vger.kernel.org
Subject: [PATCH kvm-unit-tests] vmx: do not XFAIL for virtual-APIC address
 beyond RAM
Date: Mon, 15 Apr 2019 15:10:16 +0200
Message-Id: <1555333816-17181-1-git-send-email-pbonzini@redhat.com>
X-Mailer: git-send-email 1.8.3.1
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

We will allow this behavior for KVM in some specific cases
(CR8 load/store exits enabled, virtualize APIC accesses
disabled).  Ensure these specific values of the controls
are there in the VMCS, and remove the XFAIL.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---
 x86/vmx_tests.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/x86/vmx_tests.c b/x86/vmx_tests.c
index ab7e8cc..0ca5363 100644
--- a/x86/vmx_tests.c
+++ b/x86/vmx_tests.c
@@ -3669,9 +3669,17 @@ static void test_msr_bitmap(void)
  */
 static void test_apic_virt_addr(void)
 {
+	/*
+	 * Ensure the processor will never use the virtual-APIC page, since
+	 * we will point it to invalid RAM.  Otherwise KVM is puzzled about
+	 * what we're trying to achieve and fails vmentry.
+	 */
+	u32 cpu_ctrls0 = vmcs_read(CPU_EXEC_CTRL0);
+	vmcs_write(CPU_EXEC_CTRL0, cpu_ctrls0 | CPU_CR8_LOAD | CPU_CR8_STORE);
 	test_vmcs_addr_reference(CPU_TPR_SHADOW, APIC_VIRT_ADDR,
 				 "virtual-APIC address", "Use TPR shadow",
-				 PAGE_SIZE, true, true);
+				 PAGE_SIZE, false, true);
+	vmcs_write(CPU_EXEC_CTRL0, cpu_ctrls0);
 }
 
 /*
